Make a API calls and return the response

sulthan-ahmed opened the pull request:
Using two apis axos and also the SmartSurveyClient.  

- Added the tests

sulthan-ahmed committed:
Make a API call and log the response on startup
* Use axios for the api call as it uses promises
* Add a config for api creds, baseurl, surveyID, endpoint
* Build the api from the config

sulthan-ahmed committed:
Add args for getResponses API call
- This is to be able to test it for the future

sulthan-ahmed commented on lib/api-call.js:
```diff
@@ -0,0 +1,15 @@
+
+module.exports = {
+  getResponses: function(baseUrl, surveyID, endPoint, apiToken, apiTokenSecret) {
+    axios.get(baseUrl + '/' + surveyID + '/' + endPoint + '?api_token=' + apiToken + '&api_token_secret=' + apiTokenSecret)
```
I would prefer this to be done in es6 format `${baseURL}/${surveyID}...` but it complained when attempting to do so

I would be grateful for ideas to make this look prettier

chrisns replied:
you could do with something to build the GET url query params up based on an object.
https://nodejs.org/api/querystring.html
`querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' })`

for the building the `baseUrl + '/' + surveyID + '/' + endPoint` to check you're doing: 
```
`${baseUrl}/${surveyID}/${endPoint}`
```
yeh?

daniel-ac-martin replied:
@chrisns's solution is also nice because it will fix a minor bug in your code. Specifically, anything that goes into a URL should first be escaped / URL encoded.

sulthan-ahmed replied:
good idea @chrisns 
yeah it was backtick but my sublime editor was complaining about es6 syntax.  

👍 @daniel-ac-martin

sulthan-ahmed commented on app.js:
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
+const responses = config.endPointResponses;
+const apiToken = config.apiToken;
+const apiTokenSecret = config.apiTokenSecret;
 
```
I had the above const in the api-call file itself but I found it tricky to tests so I've moved them

daniel-ac-martin replied:
Good idea to move them. I wouldn't bother setting consts here though. Unless you are using them multiple times.

Personally, I'd just do:
```js
apiCall.getResponses(
  config.baseUrl,
  config.surveyID,
  config.responses,
  config.apiToken,
  config.apiTokenSecret);
```

sulthan-ahmed replied:
nice one

chrisns commented on app.js:
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
 const hoganExpressStrict = require('hogan-express-strict');
 const expressPartialTemplates = require('express-partial-templates');
 const _ = require('lodash');
+const apiCall = require('./lib/api-call');
```
`smartSurveyAPICall` or similar would be a better name, you might have other apis

sulthan-ahmed replied:
ahh good spot @chrisns about the npm package, I should've checked for that, I'll give it a try

sulthan-ahmed replied:
fair point about the name

chrisns commented on app.js:
```diff
@@ -7,6 +7,13 @@ const govukTemplate = require('hof-govuk-template');
 const _ = require('lodash');
+const apiCall = require('./lib/api-call');
+const config = require('./config');
+const baseUrl = config.apiBaseUrl;
```
if you *have* to do this rather than refer to `config.var` then consider object destructuring

sulthan-ahmed replied:
config.var :thinking: sorry basic question time, I've googled config.var and err I didn't quite find anything except referring to variables in my config so didn't quite understand this 

object deconstructing? ahh just read up about it, I did not know that quality, you mean this right? http://wesbos.com/destructuring-objects/

chrisns replied:
sorry I meant more `config.*` like @daniel-ac-martin 's comment about
```js
apiCall.getResponses(
  config.baseUrl,
....
```

chrisns replied:
yes, thats what I meant about destructuring :)

sulthan-ahmed replied:
👍 is this also object deconstructing or it doesn't have a special name?

chrisns replied:
@daniel-ac-martin 's suggestion I don't think has a special name(?) its definitely not destructuring

chrisns commented on config.js:
```diff
@@ -0,0 +1,10 @@
+  apiToken: process.env.API_TOKEN,
+  apiTokenSecret: process.env.API_TOKEN_SECRET,
+  apiBaseUrl: process.env.BASE_URL || 'https://api.smartsurvey.io/v1/surveys',
+  surveyID: process.env.SURVEY_ID || '309701',
```
you probably don't want to hardcode that as a default survey to use

sulthan-ahmed replied:
this was pointing to the test version of the survey data so it made my life easier to have it default to that, are you saying I shouldn't have a default like that? 
The plan was to default to the live one

chrisns commented on lib/api-call.js:
```diff
@@ -0,0 +1,15 @@
+'use strict';
```
consider making this into a separate node package, or at least keeping it decoupled (as you have done) from the rendering side of things so that you can do that.

easternbloc replied:
also the file name. be more descriptive.

chrisns reviewed (changes requested):
did you look at https://www.npmjs.com/package/smartsurvey-client is it no good?

daniel-ac-martin commented on lib/api-call.js:
```diff
@@ -0,0 +1,15 @@
+        console.log(response);
+      })
+      .catch(function (error) {
+        console.log(error);
```
Would be nice to have more sophisticated logging that just console.log but perhaps it probably doesn't matter for a first iteration.

daniel-ac-martin reviewed (changes requested):
Minor potential URL encoding bug.

daniel-ac-martin commented:
I'd also suggest that the tests go in with this PR as it sounds like they pertain to this code. They shouldn't really go in as a separate PR in my opinion.

sulthan-ahmed commented:
@daniel-ac-martin I get what you're saying about including my test in this PR but I wanted to get thoughts about this PR while I work on the test.   That's already kind of proved quite handy

chrisns commented:
TDD ftw :)

easternbloc commented:
![](https://media2.giphy.com/media/26BRIaGqykNr0XKlG/giphy.gif)

sulthan-ahmed commented:
Fair, next time I'll do a WIP

sulthan-ahmed committed:
Add getResponses tests simplify code in app.js and getResponses function

sulthan-ahmed committed:
Experiment with the smartsurvey API

sulthan-ahmed committed:
Getting proxyquire and stub to work with axios

sulthan-ahmed committed:
Use smartSurveyClient api module, add api tests for axios and smartSurveyClient

sulthan-ahmed commented on lib/smartSurveyAPICall.js:
```diff
@@ -0,0 +1,36 @@
+      console.log(prettyResult);
+    }
+  }
+};
```
You might be wondering why I am using two different apis.

I am using both as I'm experimenting with the usage of both, axios is more flexible and the smartSurveyClient is missing some methods that I may need.

chrisns replied:
if you do need those methods, please make a PR to `smartSurveyClient` :)

sulthan-ahmed replied:
fair point @chrisns

sulthan-ahmed commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "start": "node .",
-    "test": "NODE_ENV=test mocha",
+    "test": "NODE_ENV=test mocha test/*",
```
strangely I have to tell mocha the location of the directory

danielepolencic replied:
*I think* this is due to the fact that you have a `lib` folder in your `test` folder. You could remove the `lib` folder or add `mocha --recursive`.

sulthan-ahmed replied:
DANIELE where you have been, haven't had you review my PRs in aaages.  `mocha --recursive` worked like a charm

So I was taught that my test folder structure should follow the same structure as my app, that's why I have a lib folder.  The idea was so that it's easy to find the related test.  Is that not necessarily true? Have you got a different thought?

sulthan-ahmed commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,100 @@
+
+// Proxyquire syntax: first argument: point to your module (which holds your dependency)
+// 2nd argument: An object with the key: module name to proxy & the value the empty object above
+// So this means smartSurveyAxiosAPI will now point to this empty object, i.e. a stub
```
These comments are intentional for learning

sulthan-ahmed commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,100 @@
+});
+
+// create an empty function using echma 5 syntax because of using the this context
+var stubSmartSurvey = function(params) {
```
getting strange behaviour when I use echma 6 syntax for context stuff

danielepolencic replied:
ECMA. Can you provide an example?

chrisns replied:
`function () {}` is still ECMA 6 syntax, it's just not an arrow function `() => {}` which as you say won't have a `this` context. and cannot be used for describing constructors which is what he's doing here.
https://www.sitepoint.com/es6-arrow-functions-new-fat-concise-syntax-javascript/
covers it reasonably well

sulthan-ahmed commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,100 @@
+          smartSurveyClientAPI.createClient.call(that, "a", "b");
+          that.client.should.be.instanceof(stubSmartSurvey);
+      })
+      it('calls the SmartSurveyClient constructor with the correct params', function() {
```
this test is failing
1. is it relevant?
2. how do I fix it?

danielepolencic replied:
1. There're few source of confusion in this test.
  - It's a good idea to test the invocation, but it's duplication to test that SmartSurveyClient behaves the way you want. Testing if the constructors receives the arguments correctly is a test that should be part of the library, not the code using it.
  - You stubbed the class and tested the invocation of the stubbed method. This test won't help you in the long run, because if the library changes the signature of the class or constructor, your test will still pass. You stubbed too much 😄 
2. I'd delete the test.

chrisns replied:
@danielepolencic 
the point of the test is to assert that calling func a(1,2) results in a call to func b({apiToken:1, apiTokenSecret:2}) so I'd say its a valid test because he's testing his code, not the library

by keeping the test he'll know that if a change does happen in the library he'll know that his code works, but integration tests will pick up the fault with the library

sulthan-ahmed commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,100 @@
+      })
+
+      it('SmartSurveyAPItest: when there is an error, log it out', function() {
+        consoleSpy = sinon.stub(console, 'log');
```
this test is not running properly because of the stub on the console

danielepolencic replied:
This issue is very similar to having proxy-require to mock smartSurveyClientAPI. There are some workarounds, but they most require to re-arrange your code (and your tests). In this particular case you could return a promise and let the user decide how they want to log the object/error.

Also, `parseResponse` is supposed to be a function, but you have a string?

sulthan-ahmed committed:
Amend smartSurveyClient getData function to take callback, add test for these, add sinon-chai, add --recursive for mocha

- Added a callback to simplify the getData function

sulthan-ahmed committed:
parseResponse now returns an err or result & tests, fixed smartSurveyClient constructor test

sulthan-ahmed commented on app.js:
```diff
@@ -21,6 +23,10 @@ app.get('/', function(req, res, next){
 })
 
+app.get('/responses', function(req, res, next){
+  res.json(smartSurveyAPICall.getData(config.apiToken, config.apiTokenSecret, config.surveyID, smartSurveyAPICall.parseResponse))
```
Separate PR questions: Do I need to just pass this to locals and then I can render it?

chrisns replied:
what does that mean?

sulthan-ahmed replied:
What I mean it's not for this PR but if I want to render the stuff from the JSON, is it simply just render

danielepolencic commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,128 @@
+        const result = parseResponse(null, 'good response');
+        const expect = JSON.stringify('good response', null, 2);
+        result.should.deep.equal(expect);
+      })
```
I'm fine with not having semicolons as long as:

- you're consistent
- you know how ASI works and all its edge cases

danielepolencic commented on lib/smartSurveyAPICall.js:
```diff
@@ -0,0 +1,36 @@
+      })
+      .catch(function (error) {
+        console.log(error);
+      });
```
This is probably very minor, buy you could use a lambda instead of an anonymous function `function () {}` => `() => {}`

chrisns replied:
in this particular scenario you could actually do `.catch(console.log)` or `console.error`

danielepolencic commented on lib/smartSurveyAPICall.js:
```diff
@@ -0,0 +1,36 @@
+module.exports = {
+  getResponses: (baseUrl, surveyID, endPoint, token, tokenSecret) => {
+    querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' });
+    const apiUrlPart = querystring.stringify({'api_token':token, 'api_token_secret': tokenSecret});
```
inconsistent spacing: `'api_token':token` should be `'api_token': token`.

Please consider using ESLint with a preset such as https://www.npmjs.com/package/eslint-config-airbnb-base to enforce consistent styling.

danielepolencic commented on test/lib/smartSurveyAPICall.test.js:
```diff
@@ -0,0 +1,128 @@
+    describe('smartSurveyClientAPI', function() {
+      var that;
+      beforeEach( function() {
+        that = {
```
It's really hard to grasp what `that` is referring to by just looking at its name. Please consider using a different name. What is `that` referring to?

daniel-ac-martin replied:
I'd also avoid 'that' as it is usually used for another purpose. (Creating a closure for a 'this' that would overwise get clobbered.)

sulthan-ahmed replied:
@danielepolencic That is referring to the context so I'll change the name to that.

@daniel-ac-martin what is that usually used for?

wilko55 commented:
Is it worth including a mock route that returns some json so you can predictably get a response? Or use something like https://github.com/typicode/json-server. Could be useful if the next stage is to manipulate the returned json

wilko55 commented on lib/smartSurveyAPICall.js:
```diff
@@ -0,0 +1,36 @@
+
+const axios = require('axios');
+const querystring = require('querystring');
+const SmartSurveyClient = require('smartsurvey-client')
```
Missing semi-colon :)

daniel-ac-martin replied:
@sulthan-ahmed: Perhaps it's worth getting a linter in there. Are the HOF guys still using eslint?

chrisns replied:
eslint ftw 👍

sulthan-ahmed replied:
Thanks guys, I've learnt my lesson.  Yes we are using eslint.

wilko55 commented on lib/smartSurveyAPICall.js:
```diff
@@ -0,0 +1,36 @@
+
+  createClient: function (token, tokenSecret) {
+    this.client = new SmartSurveyClient({ apiToken: token, apiTokenSecret: tokenSecret})
+  },
```
Inconsistent spacing here. You could go for the 
```
module.exports {
  getResponses: getResponses,
  createClient: createClient
  ...
}
```
approach which might tidy it up a little

wilko55 commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "start": "node .",
-    "test": "NODE_ENV=test mocha",
+    "test": "NODE_ENV=test mocha --recursive",
```
Do you have mocha installed globally? I had to add it to the package.json - `npm install -S -D mocha` to be able to run `npm test` on my local machine

chrisns replied:
@wilko55 npm will know the path to local mocha, doesn't need to be globally installed

wilko55 replied:
:+1: this occurred around the same time my nvm and other path variables got corrupted

chrisns commented:
I think before merging this, you should run a linter, so all the comments around that, and the next PR we review won't be mostly linting changes.
`eslint --fix` is your friend 😄

sulthan-ahmed committed:
API test: change var from that to context

This is to make it easier to understand what the variable is referring to

sulthan-ahmed committed:
Add editorconfig, eslint & fix linting issues

chrisns commented on test/lib/smart-survey-api.test.js:
```diff
@@ -1,36 +1,39 @@
+const sinon = require('sinon');
 const chai = require('chai').use(require('sinon-chai'));
+/* eslint-disable no-unused-vars*/
 const should = chai.should();
```
the linter is right, you have an unused var
you can just do `chai.should();`

chrisns commented on lib/smart-survey-api.js:
```diff
@@ -0,0 +1,36 @@
+  getResponses: (baseUrl, surveyID, endPoint, token, tokenSecret) => {
+    querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' });
+    const apiUrlPart = querystring.stringify({'api_token': token, 'api_token_secret': tokenSecret});
+    axios.get(`${baseUrl}/${surveyID}/${endPoint}?${apiUrlPart}`)
```
have you done enough exploration now to use `axios` OR `SmartSurveyClient`?

sulthan-ahmed replied:
Is there any harm keeping both? There's an api call that isn't covered for SmartSurveyClient where I need to use axios.  

I know you suggested putting in a PR for SmartSurveyClient.  That I'll have to put in as a later stage, don't you consider this PR out of scope for that?

chrisns replied:
what is the api call that is not covered by SmartSurveyClient?

sulthan-ahmed replied:
anything to do with tracking links https://docs.smartsurvey.io/v1/reference#links

chrisns replied:
if you need to use that, then you can easily either augment/extend or fork+PR the request for that and still benefit from the smartsurveyclient's pagination, auth and other abstractions, rather than having to do that yourself

sulthan-ahmed replied:
thanks @chrisns 
I'm not sure how to augment/extend it, is there any harm accepting this PR as it is and then I can do a separate PR for what you are suggesting? 
I'm just wondering if this PR is just getting bigger and bigger than what it needs to be?

chrisns replied:
you're not doing anything with `links` yet. so arguably you could make the PR **smaller** by removing the stuff you don't need *yet*

sulthan-ahmed replied:
I'm happy to remove it.  

Is there anyway, I could keep the code for future reference?

chrisns replied:
its in the git history :)

sulthan-ahmed replied:
@easternbloc has a good idea, I'll add a new commit and delete it

chrisns commented on lib/smart-survey-api.js:
```diff
@@ -0,0 +1,36 @@
+      return err;
+    }
+
+      const prettyResult = JSON.stringify(result, null, 2);
```
indentation?

chrisns commented on test/lib/.eslintrc:
```diff
@@ -0,0 +1,6 @@
+extends:
```
do you really need to put these at each directory level or is the one in `test/` enough?

sulthan-ahmed replied:
ahh good spot, I don't know how that happened

chrisns commented on test/lib/smart-survey-api.test.js:
```diff
@@ -0,0 +1,132 @@
+});
+
+// create an empty function using not using arrow functions because using context
+var stubSmartSurvey = function empty() {
```
why are you naming the function different to what it's assigned to?
this will be executable with both `stubSmartSurvey()` and `empty()` which is probably not what you intended

sulthan-ahmed replied:
yeah definitely wasn't my intention, the linter was complaining about anonymous functions so I was just giving it a name

chrisns replied:
http://eslint.org/docs/rules/func-names
the `as-needed` seems like a pragmatic option for you, looks like you've maybe got an 'always' set?

sulthan-ahmed replied:
Surprisingly, I tried that and I got more linting issues requesting not to have unnamed functions in my describe blocks and it blocks

chrisns commented on test/lib/smart-survey-api.test.js:
```diff
@@ -0,0 +1,132 @@
+      stub.restore();
+    });
+  });
+    describe('smartSurveyClientAPI', function() {
```
indentation?

chrisns commented on test/lib/smart-survey-api.test.js:
```diff
@@ -0,0 +1,132 @@
+      });
+  });
+
+      it('should call the callback with an error response when we receive an error from the api', function() {
```
indentation?

sulthan-ahmed committed:
SmartSurveyAPI: fix formating and name function to what is assigned to

sulthan-ahmed committed:
eslintrc remove duplicate files in test/lib & fix formating of smart-survey-api test

chrisns commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 node_modules
 public
 npm-debug.log
+package-lock.json
```
why do you want to ignore this?

sulthan-ahmed replied:
A lot of dependency issues that we've experienced in a ton of apps for service ops.  e.g.
https://github.com/npm/npm/issues/16866

I can remove this because I don't think this change needs to be discussed as part of this PR

sulthan-ahmed committed:
gitignore remove package-lock

sulthan-ahmed committed:
remove getResponses function using axios library & tests

This is because smartSurveyClient api is being used instead

sulthan-ahmed commented:
@chrisns has sadly left us 😢 However, I've addressed the issues he has raised. @danielepolencic has also approved the PR so I will now merge this PR 🎆
